### PR TITLE
[Medium] Initial implementation of direct global installs

### PR DESCRIPTION
**Pull request by charlespierce**
Info
-----
* First pass, naive enabling of `npm i -g <package>` to go through the `volta install` logic (using the package manager of choice) and install the package with Volta's stable platform guarantees and delegating shims.
* Completes the plumbing to make the happy path work, but doesn't handle edge cases at all.

Changes
-----
* Added `DirectInstall` type in `tool::package_global` that provides the interface necessary for modifying global installs to work with Volta's install process.
    * Specifically it exposes the ability to set up a `command` with the appropriate environment variables and to complete an install, moving files out of the `tmp` directory and into the appropriate location in the Volta data directory.
* Created a `PackageInstallCommand` that builds a command for installing a package in much the same way as `ToolCommand` does for regular commands.
    * This makes sure to call `setup_command` immediately before executing the command and `complete_install` after, if the command is successful.
* Updated `run::executor` to include an `Executor` enum that handles delegation to the appropriate underlying object (`ToolCommand` or `PackageInstallCommand`).
* Updated `npm::command` and `yarn::command` to detect a global install and return a `PackageInstallCommand` instead of a `ToolCommand`, allowing the install process to seamlessly hook into the `run` mechanisms.

Tested
-----
* Since there are still a number of outstanding edge cases (see the Notes), no new automated tests were written yet.
* Manual testing on both Unix and Windows confirms that the happy path works as expected, allowing the installed commands to be run immediately after `npm i -g` or `yarn global add`

Notes
-----
* There are a number of edge cases that still need to be handled:
    * The tool itself needs to be parsed to remove any `@<version>` specifiers, currently only the bare package name works. (Resolved in #831)
    * The Volta-managed tools like Node, Yarn, npm, etc. need to be special-cased even further to use Volta's logic for installing (with a note of what is happening, since the output will be different) (Resolved in #831)
    * We need to handle multiple installs (e.g. `npm i -g typescript ember-cli cowsay` as a single command). The plan here is to split them up and run them individually so that each gets their own sandbox (with an appropriate note to the user of what is happening).
    * We need to handle the uninstall / remove case as well, calling the `volta uninstall` logic.
* Once the above are resolved, we also need to add unit, acceptance, and smoke tests to ensure that the new flows work correctly.
* This PR builds on #827, so will remain a draft until that is merged. To see only the changes from this PR, use [this link](https://github.com/volta-cli/volta/pull/829/files/cf0dfb7517811e433269961aa65f9e0aa8cec50c..73eeca2c197db4820d8399ad34b7eb9dd9ba2305)

### CODE REVIEW

File: crates/volta-core/src/run_package_global/yarn.rs [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
+use super::executor::{Executor, PackageInstallCommand, ToolCommand, ToolKind};
+use super::{debug_active_image, debug_no_platform, CommandArg};
 use crate::error::{ErrorKind, Fallible};
 use crate::platform::{Platform, Source, System};
 use crate::session::Session;
+use crate::tool::package::PackageManager;
+
+/// Build an `Executor` for Yarn
+///
+/// If the ocmmand is a global add _and_ we have a default platform available, then we will use
```
Comment (chriskrycho) [inline crates/volta-core/src/run_package_global/yarn.rs:12]: Typo!

```suggestion
/// If the command is a global add _and_ we have a default platform available, then we will use
```

**Review by chriskrycho — CHANGES_REQUESTED** [review]
A couple small tweaks we discussed, but this is nice!

File: crates/volta-core/src/run_package_global/executor.rs [hunk 2]
```diff
@@ -85,3 +119,98 @@ impl ToolCommand {
...
+        V: AsRef<OsStr>,
+    {
+        self.command.envs(envs);
+    }
+
+    /// Updates the Platform for the command to include values from the command-line
+    pub fn cli_platform(&mut self, cli: CliPlatform) {
+        // Invariant: Platform is always set, except during the execution of `CliPlatform::merge`
+        // `CliPlatform::merge` _can't_ call methods on this object, unwrapping is safe
+        self.platform = Some(cli.merge(self.platform.take().unwrap()));
```
[the remark this answers is not in the document]
Comment (charlespierce) [reply crates/volta-core/src/run_package_global/executor.rs]: Done, and that actually does clean it up a lot, good catch!
